From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sat, 5 Oct 2019 16:07:06 +0300
Subject: [PATCH] Server specific IP and profile forwarding

Improved patch of previous version with support for per-server modern
and legacy forwarding types

diff --git a/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java b/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java
index a6aecceb7bc77b9005a3c28ba2f773929fbc5746..8b9834ea3a561624cd1aa43ace4b69f1379e8387 100644
--- a/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java
+++ b/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java
@@ -32,11 +32,6 @@ public interface FireflyProxyConfig {
      */
     boolean isLogInvalidQueryPackets();
 
-    /**
-     * @return Whether Velocity modern forwarding support is enabled or not
-     */
-    boolean isVelocityForwardingSupport();
-
     /**
      * @return Velocity modern forwarding secret
      */
diff --git a/api/src/main/java/eu/mikroskeem/firefly/api/connection/ForwardingType.java b/api/src/main/java/eu/mikroskeem/firefly/api/connection/ForwardingType.java
new file mode 100644
index 0000000000000000000000000000000000000000..e1a651748b7f3e7056775e3d3c0c5c7469d206f6
--- /dev/null
+++ b/api/src/main/java/eu/mikroskeem/firefly/api/connection/ForwardingType.java
@@ -0,0 +1,51 @@
+package eu.mikroskeem.firefly.api.connection;
+
+import java.util.Locale;
+
+/**
+ * @author Mark Vainomaa
+ */
+public enum ForwardingType {
+    /**
+     * IP and profile forwarding is off (vanilla)
+     */
+    OFF,
+
+    /**
+     * IP and profile forwarding is done in BungeeCord way by modifying handshake packet
+     */
+    BUNGEECORD,
+
+    /**
+     * IP and profile forwarding is done in Velocity way by utilizing Login Plugin Request/Response packets,
+     * which are available in 1.13+
+     */
+    VELOCITY
+    ;
+
+    public boolean isOff() {
+        return this == OFF;
+    }
+
+    public boolean isVelocity() {
+        return this == VELOCITY;
+    }
+
+    public boolean isBungeeCord() {
+        return this == BUNGEECORD;
+    }
+
+    public static ForwardingType fromString(String rawValue) {
+        if (rawValue.equalsIgnoreCase("true") || rawValue.equalsIgnoreCase("false")) {
+            // Legacy
+            return rawValue.equalsIgnoreCase("true") ? BUNGEECORD : OFF;
+        }
+
+        rawValue = rawValue.toUpperCase(Locale.ROOT);
+        try {
+            return valueOf(rawValue);
+        } catch (IllegalArgumentException e) {
+            throw new IllegalArgumentException("Invalid forwarding type: " + rawValue);
+        }
+    }
+}
diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
index d002a7445e01b3bd5d241b00e819a888acfa4a54..4fb880e6915467cea4ec21ba00c7782aa854f824 100644
--- a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
+++ b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
@@ -191,6 +191,9 @@ public interface ProxyConfig extends eu.mikroskeem.firefly.api.config.FireflyPro
      */
     @Deprecated
     boolean isIpForward();
+    // Firefly start - server specific IP forwarding
+    eu.mikroskeem.firefly.api.connection.ForwardingType getIpForwardType();
+    // Firefly end
 
     /**
      * The encoded favicon.
diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyServer.java b/api/src/main/java/net/md_5/bungee/api/ProxyServer.java
index 54a1e5094ea5212e86be26832389d427824153d9..e7550b00755aa6259a66650d5e18a1d9f8399e77 100644
--- a/api/src/main/java/net/md_5/bungee/api/ProxyServer.java
+++ b/api/src/main/java/net/md_5/bungee/api/ProxyServer.java
@@ -242,8 +242,13 @@ public abstract class ProxyServer
      * @param restricted whether the server info restricted property will be set
      * @return the constructed instance
      */
-    public abstract ServerInfo constructServerInfo(String name, InetSocketAddress address, String motd, boolean restricted);
+    // Firefly start - server specific IP forwarding
+    public ServerInfo constructServerInfo(String name, InetSocketAddress address, String motd, boolean restricted) {
+        return constructServerInfo(name, (SocketAddress) address, motd, restricted, getConfig().getIpForwardType());
+    }
+    // Firefly end
 
+    // Firefly start - server specific IP forwarding
     /**
      * Factory method to construct an implementation specific server info
      * instance.
@@ -252,9 +257,44 @@ public abstract class ProxyServer
      * @param address connectable Minecraft address + port of the server
      * @param motd the motd when used as a forced server
      * @param restricted whether the server info restricted property will be set
+     * @param ipForwardType IP forwarding method used by this server
      * @return the constructed instance
      */
-    public abstract ServerInfo constructServerInfo(String name, SocketAddress address, String motd, boolean restricted);
+    public ServerInfo constructServerInfo(String name, InetSocketAddress address, String motd, boolean restricted, eu.mikroskeem.firefly.api.connection.ForwardingType ipForwardType) {
+        return constructServerInfo(name, (SocketAddress) address, motd, restricted, ipForwardType);
+    }
+
+    // Firefly start - server specific IP forwarding
+    /**
+     * Factory method to construct an implementation specific server info
+     * instance.
+     *
+     * @param name name of the server
+     * @param address connectable Minecraft address + port of the server
+     * @param motd the motd when used as a forced server
+     * @param restricted whether the server info restricted property will be set
+     * @param ipForwardType IP forwarding method used by this server
+     * @return the constructed instance
+     */
+    public abstract ServerInfo constructServerInfo(String name, SocketAddress address, String motd, boolean restricted, eu.mikroskeem.firefly.api.connection.ForwardingType ipForwardType);
+    // Firefly end
+
+
+    /**
+     * Factory method to construct an implementation specific server info
+     * instance.
+     *
+     * @param name name of the server
+     * @param address connectable Minecraft address + port of the server
+     * @param motd the motd when used as a forced server
+     * @param restricted whether the server info restricted property will be set
+     * @return the constructed instance
+     */
+    // Firefly start - server specific IP forwarding
+    public ServerInfo constructServerInfo(String name, SocketAddress address, String motd, boolean restricted) {
+        return constructServerInfo(name, address, motd, restricted, getConfig().getIpForwardType());
+    }
+    // Firefly end
 
     /**
      * Returns the console overlord for this proxy. Being the console, this
diff --git a/api/src/main/java/net/md_5/bungee/api/config/ServerInfo.java b/api/src/main/java/net/md_5/bungee/api/config/ServerInfo.java
index b0fc000a32f35d68623fdf07f173e534f30ae42d..c9e274e7883d69405137cb668713fc25009224bd 100644
--- a/api/src/main/java/net/md_5/bungee/api/config/ServerInfo.java
+++ b/api/src/main/java/net/md_5/bungee/api/config/ServerInfo.java
@@ -115,4 +115,12 @@ public interface ServerInfo
      * @param callback the callback to call when the count has been retrieved.
      */
     void ping(Callback<ServerPing> callback);
+    // Firefly start - server specific IP forwarding
+    /**
+     * Returns IP and profile forwarding method used by this server
+     *
+     * @return IP and profile forwarding method used by this server
+     */
+    default eu.mikroskeem.firefly.api.connection.ForwardingType getIpForwardType() { return net.md_5.bungee.api.ProxyServer.getInstance().getConfig().getIpForwardType(); }
+    // Firefly end
 }
diff --git a/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java b/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java
index d08986932da4bc4a704feb4de9aeeeeb5bffd638..48d177aa48f9a756aa44aac7ddf75006e355606e 100644
--- a/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java
+++ b/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java
@@ -37,9 +37,6 @@ public class FireflyConfiguration extends WaterfallConfiguration {
     @Getter
     private boolean logInvalidQueryPackets = false;
 
-    @Getter
-    private boolean velocityForwardingSupport = false;
-
     @Getter
     private byte[] velocityForwardingSecret;
 
@@ -83,7 +80,6 @@ public class FireflyConfiguration extends WaterfallConfiguration {
         restartScriptPath = config.getString("restart.script_path", restartScriptPath);
         ignoredLogCommands = new HashSet<>(config.getList("log.ignored_commands", () -> ImmutableList.of("login", "register", "changepassword")));
         logInvalidQueryPackets = config.getBoolean("log.invalid_query_packets", logInvalidQueryPackets);
-        velocityForwardingSupport = config.getBoolean("velocity_modern_forwarding.enabled", velocityForwardingSupport);
         velocityForwardingSecret = config.getString("velocity_modern_forwarding.secret", () -> VelocitySupport.generateRandomString(12)).getBytes(StandardCharsets.UTF_8);
         tcpFastOpenEnabled = config.getBoolean("networking.tcp_fast_open.enabled", tcpFastOpenEnabled);
         tcpFastOpenMode = setupTfo(config.getInt("networking.tcp_fast_open.mode", tcpFastOpenMode.value));
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 6c5f91a62179948508b028999f28fb82e141e58a..3ee64e9982bb09dfa08ccea4456950fc90d06fb6 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -783,15 +783,10 @@ public class BungeeCord extends ProxyServer
     }
 
     @Override
-    public ServerInfo constructServerInfo(String name, InetSocketAddress address, String motd, boolean restricted)
+    public ServerInfo constructServerInfo(String name, SocketAddress address, String motd, boolean restricted, eu.mikroskeem.firefly.api.connection.ForwardingType ipForwardType) // Firefly - server specific IP forwarding
     {
-        return constructServerInfo( name, (SocketAddress) address, motd, restricted );
-    }
-
-    @Override
-    public ServerInfo constructServerInfo(String name, SocketAddress address, String motd, boolean restricted)
-    {
-        return new BungeeServerInfo( name, address, motd, restricted );
+        Preconditions.checkNotNull( ipForwardType, "ipForwardType" ); // Firefly - server specific IP forwarding
+        return new BungeeServerInfo( name, address, motd, restricted, ipForwardType ); // Firefly - server specific IP forwarding
     }
 
     @Override
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeServerInfo.java b/proxy/src/main/java/net/md_5/bungee/BungeeServerInfo.java
index 7b002089ef9dd9f365dc0f7db076976b8dc6260c..8d10c9188f642c97451ae34b44745660004e36d4 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeServerInfo.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeServerInfo.java
@@ -36,6 +36,7 @@ import net.md_5.bungee.protocol.packet.PluginMessage;
 @ToString(of =
 {
     "name", "socketAddress", "restricted"
+        , "ipForwardType" // Firefly - server specific IP forwarding
 })
 // CHECKSTYLE:ON
 public class BungeeServerInfo implements ServerInfo
@@ -52,6 +53,10 @@ public class BungeeServerInfo implements ServerInfo
     private final boolean restricted;
     @Getter
     private final Queue<DefinedPacket> packetQueue = new LinkedList<>();
+    // Firefly start - server specific IP forwarding
+    @Getter
+    private final eu.mikroskeem.firefly.api.connection.ForwardingType ipForwardType;
+    // Firefly end
 
     @Synchronized("players")
     public void addPlayer(ProxiedPlayer player)
diff --git a/proxy/src/main/java/net/md_5/bungee/ServerConnector.java b/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
index 18c7daa408cb6dc990201ae7bb5ec8a3319e465a..37305a7d0523c259419e0d3cf8bb3d5081b564b0 100644
--- a/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
+++ b/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
@@ -102,7 +102,7 @@ public class ServerConnector extends PacketHandler implements eu.mikroskeem.fire
         Handshake originalHandshake = user.getPendingConnection().getHandshake();
         Handshake copiedHandshake = new Handshake( originalHandshake.getProtocolVersion(), originalHandshake.getHost(), originalHandshake.getPort(), 2 );
 
-        if ( BungeeCord.getInstance().config.isIpForward() && user.getSocketAddress() instanceof InetSocketAddress )
+        if ( target.getIpForwardType().isBungeeCord() && user.getSocketAddress() instanceof InetSocketAddress ) // Firefly - server specific IP forwarding
         {
             String newHost = copiedHandshake.getHost() + "\00" + user.getAddress().getHostString() + "\00" + user.getUUID();
 
@@ -171,7 +171,7 @@ public class ServerConnector extends PacketHandler implements eu.mikroskeem.fire
         ch.setProtocol( Protocol.GAME );
         thisState = State.LOGIN;
         // Firefly start - Implement Velocity modern forwarding support
-        if (bungee.getConfig().isVelocityForwardingSupport() && !modernInformationForwarded) {
+        if (target.getIpForwardType().isVelocity() && !modernInformationForwarded) {
             throw new QuietException("Your server did not send a forwarding request to the proxy. Is it set up correctly?");
         }
         // Firefly end
@@ -463,7 +463,7 @@ public class ServerConnector extends PacketHandler implements eu.mikroskeem.fire
     // Firefly start - Implement Velocity modern forwarding support
     @Override
     public void handle(net.md_5.bungee.protocol.packet.LoginPayloadRequest request) {
-        if (bungee.getConfig().isVelocityForwardingSupport() && request.getChannel().equals(eu.mikroskeem.firefly.misc.VelocitySupport.MODERN_FORWARDING_CHANNEL_NAME)) {
+        if (target.getIpForwardType().isVelocity() && request.getChannel().equals(eu.mikroskeem.firefly.misc.VelocitySupport.MODERN_FORWARDING_CHANNEL_NAME)) {
             ByteBuf forwardingData = null;
             try {
                 // Create forwarding data
diff --git a/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java b/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
index 89f2b533bc55219843bbd3eb9ed75ee677be82b2..52b5f8101b4541664455ce755571e751b65f3a3c 100644
--- a/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
+++ b/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
@@ -66,6 +66,7 @@ public abstract class Configuration implements ProxyConfig
     private int throttle = 4000;
     private int throttleLimit = 3;
     private boolean ipForward = true; // Firefly - sane configuration defaults
+    private eu.mikroskeem.firefly.api.connection.ForwardingType ipForwardType = eu.mikroskeem.firefly.api.connection.ForwardingType.BUNGEECORD;
     private Favicon favicon;
     private int compressionThreshold = 256;
     private boolean preventProxyConnections;
@@ -101,7 +102,18 @@ public abstract class Configuration implements ProxyConfig
         remotePingTimeout = adapter.getInt( "remote_ping_timeout", remotePingTimeout );
         throttle = adapter.getInt( "connection_throttle", throttle );
         throttleLimit = adapter.getInt( "connection_throttle_limit", throttleLimit );
+        /* // Firefly start - server specific IP forwarding
         ipForward = adapter.getBoolean( "ip_forward", ipForward );
+        */
+        Object ipForwardValue;
+        try {
+            ipForwardValue = adapter.getString("ip_forward", ipForwardType.name().toLowerCase(java.util.Locale.ROOT));
+        } catch (ClassCastException e) {
+            ipForwardValue = adapter.getBoolean("ip_forward", !ipForwardType.isOff());
+        }
+        ipForwardType = eu.mikroskeem.firefly.api.connection.ForwardingType.fromString(ipForwardValue.toString());
+        ipForward = !ipForwardType.isOff(); // Legacy
+        // Firefly end
         compressionThreshold = adapter.getInt( "network_compression_threshold", compressionThreshold );
         preventProxyConnections = adapter.getBoolean( "prevent_proxy_connections", preventProxyConnections );
         forgeSupport = adapter.getBoolean( "forge_support", forgeSupport );
diff --git a/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java b/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
index f447a9ed70c14b78daf179d545889dc87af6c49a..da7f0cde52fd7cff87251262f54ad92a95720316 100644
--- a/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
+++ b/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
@@ -233,8 +233,17 @@ public class YamlConfig implements ConfigurationAdapter
             String addr = get( "address", "localhost:25565", val );
             String motd = ChatColor.translateAlternateColorCodes( '&', get( "motd", "&1Just another Waterfall - Forced Host", val ) );
             boolean restricted = get( "restricted", false, val );
+            // Firefly start - server specific IP forwarding
+            eu.mikroskeem.firefly.api.connection.ForwardingType ipForwardingType;
+            var ipForwardingTypeRaw = get("ip_forward", (Object) ProxyServer.getInstance().getConfig().getIpForwardType().name().toLowerCase(Locale.ROOT), val).toString();
+            try {
+                ipForwardingType = eu.mikroskeem.firefly.api.connection.ForwardingType.fromString(ipForwardingTypeRaw);
+            } catch (IllegalArgumentException e) {
+                throw new IllegalArgumentException("Invalid forwarding type: " + ipForwardingTypeRaw);
+            }
+            // Firefly end
             SocketAddress address = Util.getAddr( addr );
-            ServerInfo info = ProxyServer.getInstance().constructServerInfo( name, address, motd, restricted );
+            ServerInfo info = ProxyServer.getInstance().constructServerInfo( name, address, motd, restricted, ipForwardingType ); // Firefly - server specific IP forwarding
             ret.put( name, info );
         }
 
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index 1a33bb8949558ff1b06fc84478cb89caa24ca4aa..dec1a266655251399e572e5bd2091c6ec7e589de 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -139,7 +139,7 @@ public class DownstreamBridge extends PacketHandler
     @Override
     public void handle(PacketWrapper packet) throws Exception
     {
-        con.getEntityRewrite().rewriteClientbound( packet.buf, con.getServerEntityId(), con.getClientEntityId(), con.getPendingConnection().getVersion() );
+        con.getEntityRewrite().rewriteClientbound( packet.buf, con.getServerEntityId(), con.getClientEntityId(), con.getPendingConnection().getVersion(), con.getServer() ); // Firefly - server specific IP forwarding
         con.sendPacket( packet );
     }
 
@@ -156,7 +156,7 @@ public class DownstreamBridge extends PacketHandler
     @Override
     public void handle(PlayerListItem playerList) throws Exception
     {
-        if (!this.bungee.getConfig().isAllowTablistRewrite() && con.getPendingConnection().isOnlineMode() && (this.bungee.getConfig().isVelocityForwardingSupport() || this.server.getInfo().isIpForward())) { con.getTabListHandler().onUpdate(playerList); return; } // Firefly - don't rewrite tablist on online mode connections
+        if (!this.bungee.getConfig().isAllowTablistRewrite() && con.getPendingConnection().isOnlineMode() && !this.server.getInfo().getIpForwardType().isOff()) { con.getTabListHandler().onUpdate(playerList); return; } // Firefly - don't rewrite tablist on online mode connections
         con.getTabListHandler().onUpdate( TabList.rewrite( playerList ) );
         throw CancelSendSignal.INSTANCE; // Always throw because of profile rewriting
     }
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
index f987995e43c12462977b12074ec9c03bd3323e95..6b7f4e2184c4aacc2fd29e789133b9ef4b28809b 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
@@ -118,7 +118,7 @@ public class UpstreamBridge extends PacketHandler
     {
         if ( con.getServer() != null )
         {
-            con.getEntityRewrite().rewriteServerbound( packet.buf, con.getClientEntityId(), con.getServerEntityId(), con.getPendingConnection().getVersion() );
+            con.getEntityRewrite().rewriteServerbound( packet.buf, con.getClientEntityId(), con.getServerEntityId(), con.getPendingConnection().getVersion(), con.getServer() ); // Firefly - server specific IP forwarding
             con.getServer().getCh().write( packet );
         }
     }
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
index e6afb98db1822d26d239bdfdfc9510fdb220f551..36bb08fc3ab2a061a09279eb9f295ae45c4f3f9f 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
@@ -90,24 +90,28 @@ public abstract class EntityMap
         }
     }
 
-    public void rewriteServerbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteServerbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
         rewrite( packet, oldId, newId, serverboundInts, serverboundVarInts );
     }
 
-    public void rewriteServerbound(ByteBuf packet, int oldId, int newId, int protocolVersion)
+    // Firefly start - server specific IP forwarding
+    public void rewriteServerbound(ByteBuf packet, int oldId, int newId, int protocolVersion, net.md_5.bungee.api.connection.Server server)
     {
-        rewriteServerbound( packet, oldId, newId );
+        rewriteServerbound( packet, oldId, newId, server );
+        // Firefly end
     }
 
-    public void rewriteClientbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
         rewrite( packet, oldId, newId, clientboundInts, clientboundVarInts );
     }
 
-    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, int protocolVersion)
+    // Firefly start - server specific IP forwarding
+    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, int protocolVersion, net.md_5.bungee.api.connection.Server server)
     {
-        rewriteClientbound( packet, oldId, newId );
+        rewriteClientbound( packet, oldId, newId, server );
+        // Firefly end
     }
 
     protected static void rewriteInt(ByteBuf packet, int oldId, int newId, int offset)
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_10.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_10.java
index 6db530c3aeda7481b2ef967e598f3fee5b7f5acd..0dff7bf51d770c8b8ce5e79d93f1b85d511684df 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_10.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_10.java
@@ -48,9 +48,9 @@ class EntityMap_1_10 extends EntityMap
 
     @Override
     @SuppressFBWarnings("DLS_DEAD_LOCAL_STORE")
-    public void rewriteClientbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteClientbound( packet, oldId, newId );
+        super.rewriteClientbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
 
         // Special cases
         int readerIndex = packet.readerIndex();
@@ -156,15 +156,15 @@ class EntityMap_1_10 extends EntityMap
     }
 
     @Override
-    public void rewriteServerbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteServerbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteServerbound( packet, oldId, newId );
+        super.rewriteServerbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
         // Special cases
         int readerIndex = packet.readerIndex();
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isIpForward() )
+        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ && server.getInfo().getIpForwardType().isOff() ) // Firefly - server specific IP forwarding
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_11.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_11.java
index 368221275097dd58dbfbb5b150875424570e17c5..19c3ea27cddf80ee6b7621112c9eca900a5467cf 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_11.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_11.java
@@ -48,9 +48,9 @@ class EntityMap_1_11 extends EntityMap
 
     @Override
     @SuppressFBWarnings("DLS_DEAD_LOCAL_STORE")
-    public void rewriteClientbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteClientbound( packet, oldId, newId );
+        super.rewriteClientbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
 
         // Special cases
         int readerIndex = packet.readerIndex();
@@ -157,15 +157,15 @@ class EntityMap_1_11 extends EntityMap
     }
 
     @Override
-    public void rewriteServerbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteServerbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteServerbound( packet, oldId, newId );
+        super.rewriteServerbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
         // Special cases
         int readerIndex = packet.readerIndex();
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isIpForward() )
+        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ && server.getInfo().getIpForwardType().isOff() ) // Firefly - server specific IP forwarding
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12.java
index 38e12ce400de9dab0e90145f0e68f2ce694886ba..a122e03b673e4d3d7fcdea84b941e865eec57e8b 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12.java
@@ -48,9 +48,9 @@ class EntityMap_1_12 extends EntityMap
 
     @Override
     @SuppressFBWarnings("DLS_DEAD_LOCAL_STORE")
-    public void rewriteClientbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteClientbound( packet, oldId, newId );
+        super.rewriteClientbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
 
         // Special cases
         int readerIndex = packet.readerIndex();
@@ -157,15 +157,15 @@ class EntityMap_1_12 extends EntityMap
     }
 
     @Override
-    public void rewriteServerbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteServerbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteServerbound( packet, oldId, newId );
+        super.rewriteServerbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
         // Special cases
         int readerIndex = packet.readerIndex();
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1E /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isIpForward() )
+        if ( packetId == 0x1E /* Spectate : PacketPlayInSpectate */ && server.getInfo().getIpForwardType().isOff() ) // Firefly - server specific IP forwarding
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12_1.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12_1.java
index 5f296839480bc0ec68c80b1bfca87ebf36df18ab..75e23b5ed8ae6a7a1651a763f022c04b7adcce27 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12_1.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12_1.java
@@ -48,9 +48,9 @@ class EntityMap_1_12_1 extends EntityMap
 
     @Override
     @SuppressFBWarnings("DLS_DEAD_LOCAL_STORE")
-    public void rewriteClientbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteClientbound( packet, oldId, newId );
+        super.rewriteClientbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
 
         // Special cases
         int readerIndex = packet.readerIndex();
@@ -157,15 +157,15 @@ class EntityMap_1_12_1 extends EntityMap
     }
 
     @Override
-    public void rewriteServerbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteServerbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteServerbound( packet, oldId, newId );
+        super.rewriteServerbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
         // Special cases
         int readerIndex = packet.readerIndex();
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1E /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isIpForward() )
+        if ( packetId == 0x1E /* Spectate : PacketPlayInSpectate */ && server.getInfo().getIpForwardType().isOff() ) // Firefly - server specific IP forwarding
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_13.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_13.java
index f3372f0885073469672ec1e1da379066e00d70a4..3669f09ae637da5e5f182f932da85ad40e3b9b6f 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_13.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_13.java
@@ -48,9 +48,9 @@ class EntityMap_1_13 extends EntityMap
 
     @Override
     @SuppressFBWarnings("DLS_DEAD_LOCAL_STORE")
-    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, int protocolVersion)
+    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, int protocolVersion, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteClientbound( packet, oldId, newId );
+        super.rewriteClientbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
 
         // Special cases
         int readerIndex = packet.readerIndex();
@@ -157,15 +157,15 @@ class EntityMap_1_13 extends EntityMap
     }
 
     @Override
-    public void rewriteServerbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteServerbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteServerbound( packet, oldId, newId );
+        super.rewriteServerbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
         // Special cases
         int readerIndex = packet.readerIndex();
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x28 /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isIpForward() )
+        if ( packetId == 0x28 /* Spectate : PacketPlayInSpectate */ && server.getInfo().getIpForwardType().isOff() ) // Firefly - server specific IP forwarding
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_14.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_14.java
index 8210b0a7e347e381900b5a7b2ac436c8aa085c9d..b2dc5f676effb20c0d336878c773edb421483620 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_14.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_14.java
@@ -47,9 +47,9 @@ class EntityMap_1_14 extends EntityMap
 
     @Override
     @SuppressFBWarnings("DLS_DEAD_LOCAL_STORE")
-    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, int protocolVersion)
+    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, int protocolVersion, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteClientbound( packet, oldId, newId );
+        super.rewriteClientbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
 
         // Special cases
         int readerIndex = packet.readerIndex();
@@ -161,15 +161,15 @@ class EntityMap_1_14 extends EntityMap
     }
 
     @Override
-    public void rewriteServerbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteServerbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteServerbound( packet, oldId, newId );
+        super.rewriteServerbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
         // Special cases
         int readerIndex = packet.readerIndex();
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x2B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isIpForward() )
+        if ( packetId == 0x2B /* Spectate : PacketPlayInSpectate */ && server.getInfo().getIpForwardType().isOff() ) // Firefly - server specific IP forwarding
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_15.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_15.java
index c2cf810f348462b3470da2040fd1102861a20005..b40a378045c9f7e3a5cef2ee83db3e39e26a0c15 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_15.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_15.java
@@ -47,9 +47,9 @@ class EntityMap_1_15 extends EntityMap
 
     @Override
     @SuppressFBWarnings("DLS_DEAD_LOCAL_STORE")
-    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, int protocolVersion)
+    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, int protocolVersion, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteClientbound( packet, oldId, newId );
+        super.rewriteClientbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
 
         // Special cases
         int readerIndex = packet.readerIndex();
@@ -161,15 +161,15 @@ class EntityMap_1_15 extends EntityMap
     }
 
     @Override
-    public void rewriteServerbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteServerbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteServerbound( packet, oldId, newId );
+        super.rewriteServerbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
         // Special cases
         int readerIndex = packet.readerIndex();
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x2B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isIpForward() )
+        if ( packetId == 0x2B /* Spectate : PacketPlayInSpectate */ && server.getInfo().getIpForwardType().isOff() ) // Firefly - server specific IP forwarding
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_8.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_8.java
index 8e2dbe69929a67af592eb6aebb8049e8cb8e6171..94835db386ed23a882d58390f02d3c7684ff3168 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_8.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_8.java
@@ -49,9 +49,9 @@ class EntityMap_1_8 extends EntityMap
 
     @Override
     @SuppressFBWarnings("DLS_DEAD_LOCAL_STORE")
-    public void rewriteClientbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteClientbound( packet, oldId, newId );
+        super.rewriteClientbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
 
         //Special cases
         int readerIndex = packet.readerIndex();
@@ -150,15 +150,15 @@ class EntityMap_1_8 extends EntityMap
     }
 
     @Override
-    public void rewriteServerbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteServerbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteServerbound( packet, oldId, newId );
+        super.rewriteServerbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
         //Special cases
         int readerIndex = packet.readerIndex();
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x18 /* Spectate */ && !BungeeCord.getInstance().getConfig().isIpForward() )
+        if ( packetId == 0x18 /* Spectate */ && server.getInfo().getIpForwardType().isOff() ) // Firefly - server specific IP forwarding
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9.java
index d61dc0cbaf00b931523a9c3bbdc493e362d854e1..5698d1d44f01f7f418ffc6dde041b27f0feb0f1a 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9.java
@@ -48,9 +48,9 @@ class EntityMap_1_9 extends EntityMap
 
     @Override
     @SuppressFBWarnings("DLS_DEAD_LOCAL_STORE")
-    public void rewriteClientbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteClientbound( packet, oldId, newId );
+        super.rewriteClientbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
 
         // Special cases
         int readerIndex = packet.readerIndex();
@@ -156,15 +156,15 @@ class EntityMap_1_9 extends EntityMap
     }
 
     @Override
-    public void rewriteServerbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteServerbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteServerbound( packet, oldId, newId );
+        super.rewriteServerbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
         // Special cases
         int readerIndex = packet.readerIndex();
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isIpForward() )
+        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ && server.getInfo().getIpForwardType().isOff() ) // Firefly - server specific IP forwarding
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9_4.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9_4.java
index bfbc84327c8d590f1fe1c47c83b63d9975394f54..071af648d8ae8065baaaed2af63430cb1f69a7fb 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9_4.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9_4.java
@@ -48,9 +48,9 @@ class EntityMap_1_9_4 extends EntityMap
 
     @Override
     @SuppressFBWarnings("DLS_DEAD_LOCAL_STORE")
-    public void rewriteClientbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteClientbound( packet, oldId, newId );
+        super.rewriteClientbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
 
         // Special cases
         int readerIndex = packet.readerIndex();
@@ -156,15 +156,15 @@ class EntityMap_1_9_4 extends EntityMap
     }
 
     @Override
-    public void rewriteServerbound(ByteBuf packet, int oldId, int newId)
+    public void rewriteServerbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) // Firefly - server specific IP forwarding
     {
-        super.rewriteServerbound( packet, oldId, newId );
+        super.rewriteServerbound( packet, oldId, newId, server ); // Firefly - server specific IP forwarding
         // Special cases
         int readerIndex = packet.readerIndex();
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isIpForward() )
+        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ && server.getInfo().getIpForwardType().isOff() ) // Firefly - server specific IP forwarding
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_Dummy.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_Dummy.java
index cb81d1dd8f361c1845cace634c7bd22a9e6413b3..53a862c8070d259728e6146bb05c90b1758e2203 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_Dummy.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_Dummy.java
@@ -12,19 +12,19 @@ public class EntityMap_Dummy extends EntityMap {
     }
 
     @Override
-    public void rewriteServerbound(ByteBuf packet, int oldId, int newId) {
+    public void rewriteServerbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) { // Firefly - server specific IP forwarding
     }
 
     @Override
-    public void rewriteServerbound(ByteBuf packet, int oldId, int newId, int protocolVersion) {
+    public void rewriteServerbound(ByteBuf packet, int oldId, int newId, int protocolVersion, net.md_5.bungee.api.connection.Server server) { // Firefly - server specific IP forwarding
     }
 
     @Override
-    public void rewriteClientbound(ByteBuf packet, int oldId, int newId) {
+    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, net.md_5.bungee.api.connection.Server server) { // Firefly - server specific IP forwarding
     }
 
     @Override
-    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, int protocolVersion) {
+    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, int protocolVersion, net.md_5.bungee.api.connection.Server server) { // Firefly - server specific IP forwarding
     }
 }
 // Waterfall end
\ No newline at end of file
